bundle/edu.gemini.sp.vcs/src/main/scala/edu/gemini/sp/vcs/diff/MergePlan.scala

```diff
@@ -1,14 +1,112 @@
 package edu.gemini.sp.vcs.diff
 
-import edu.gemini.pot.sp.ISPProgram
+import edu.gemini.pot.sp._
+import edu.gemini.pot.sp.version.{VersionMap, NodeVersions}
+import edu.gemini.sp.vcs.diff.NodeDetail.Obs
+import edu.gemini.sp.vcs.diff.VcsFailure._
+import edu.gemini.spModel.rich.pot.sp._
 
 import scalaz._
+import Scalaz._
 
 /** Describes the modifications required for a local program to complete a
   * merge.
   */
 case class MergePlan(update: Tree[MergeNode], delete: Set[Missing]) {
 
+  /** "Encode" for serialization. The issue is that `scalaz.Tree` is not
+    * `Serializable` but we need to send `MergePlan`s over `trpc`.
+    */
+  def encode: MergePlan.Transport = {
+    def encodeTree(t: Tree[MergeNode]): MergePlan.TreeTransport =
+      MergePlan.TreeTransport(t.rootLabel, t.subForest.toList.map(encodeTree))
+
+    MergePlan.Transport(encodeTree(update), delete)
+  }
+
+  /** Gets the `VersionMap` of the provided program as it will be after the
+    * updates in this plan have been applied. */
+  def vm(p: ISPProgram): VersionMap = {
+    // Extract the updates to the VersionMap from the MergePlan.
+    val vmUpdates: VersionMap = {
+      val vm0 = update.foldRight(Map.empty[SPNodeKey, NodeVersions]) { (mn, m) =>
+        mn match {
+          case Modified(k, nv, _, _) => m.updated(k, nv)
+          case _                     => m
+        }
+      }
+      (vm0/:delete) { case (vm1, Missing(k, nv)) => vm1.updated(k, nv) }
+    }
+
+    p.getVersions ++ vmUpdates
+  }
+
   /** Accepts a program and edits it according to this merge plan. */
-  def merge(p: ISPProgram): Unit = ???
+  def merge(f: ISPFactory, p: ISPProgram): TryVcs[Unit] = {
+    // Tries to create an ISPNode from the information in the MergeNode.
+    def create(mn: MergeNode): TryVcs[ISPNode] =
+      mn match {
+        case Modified(k, _, dob, _) =>
+          NodeFactory.mkNode(f, p, dob.getType, Some(k)).toRightDisjunction {
+            Unexpected("Could not create science program node of type: " + dob.getType)
+          }
+        case Unmodified(k)          =>
+          Unexpected(s"Unmodified node with key $k not found in program ${p.getProgramID}.").left
+      }
+
+    // Edit the ISPNode, applying the changes in the MergeNode if any.
+    def edit(t: Tree[(MergeNode, ISPNode)]): Unit = {
```

Comment:
Yes.  It's walking the tree mutating the `ISPNode` according to the matching `MergeNode`.